### DISCUSSION TITLE
Add hooks for adding new image transform modes by plugins

### DISCUSSION THREAD

**Discussion by piotrpog** [craftcms/cms#8944] [Ideas] [answered]
I was thinking of creating plugin that adds "letterbox" mode based on Imager functionality, but already existing transform modes (stretch, crop, fit) are pretty much hardcoded in source code of Craft.

**Comment by JshGrn** [comment]
Why is this mode not by default? Lol

**Reply by brandonkelly** [reply]
What’s the use case? Seems awkward to me; you could easily create a letterbox effect using a `fit` transform plus a dark background color in the thumbnail’s container.

**Reply by JshGrn** [reply]
It would be good for where it isn't ideal to have a container or where I will link to the image and require it to be a specific size, it does make things like displaying entries thumbnails on a archive page much, much easier without having the CSS required to letterbox effect it.

I have done that now, but still its not ideal where it could be a function, obviously the use-case is there as Imager added it and obviously have quite a few people using the plugin for that alone. Would be great to have in craft. To be honest, the way fit is described I thought it would be 'letterboxed' anyway, maybe a checkbox toggling the functionality would be good. e.g. hard limits on the width and height, and fit inside this rather than upscale as much as possible until a width or height is hit.

**Reply by brandonkelly** [reply]
Idk, I’m not convinced… saving a tiny bit of CSS work feels not worth the image filesize increase.

**Reply by JshGrn** [reply]
Surely it would be good to give the choice to the developer? It is obviously a thing that people want as every thread I find on SO or various other places on the web all point towards using Imager's letterbox mode, surely it would be in the interest of Craft to offer features which people want, given how easily this is probably implemented.

I feel this is very opinionated and therefore wont be something that finds itself coming soon, but it would be VERY good to have it without having to pay $99 USD _just for this single feature._

We have a multitude of scenarios which ultimately end up with a MUCH better result adding ~1kb to the image with server processed resizing with backgrounds. Given so many people have access to gigabit speeds its frustrating to be worried about a couple of kb.

I am really trying to migrate _hundreds_ of WP sites over to Craft as it is _much_ better, yet the matrix field/acf issue with not having conditionally showing fields coupled with simple things like this are what stops us. I really look forward to Craft 4 where I believe this will be addressed and I _really hope_ this would be a feature. Even with a warning to developers saying this can be done in CSS and would save 2kb of data.

**Comment by piotrpog** [comment]
@brandonkelly just to add on that issue.

I dont think that it is really about letterbox mode specifically, but about adding possibility to implement any other various image transform modes that developers might want.

Imager for example had three additional transform modes.

So it would be nice if you reconsidered adding such event, as i think it wouldnt be hard thing to do, just modify`_createTransformForAsset` method slightly.

**Comment by piotrpog** [comment]
Since it was 3 years since last comment to this request, i would like to bump it again. 

One of my clients recently wanted to add some custom transform mode to his website which would detect uniform color background to the sides of image and crop it automatically. 
Adding this functionality as addition to existing transform modes would simplify the whole thing.
Could you please consider it in the upcoming Craft versions?

**Comment by brandonkelly** [comment, marked as answer]
We ended up adding a Letterbox transform mode in Craft 4.4.0.

I’m going to close this now because we need to be mindful about what transform options can be added, since PHP (GD/ImageMagick) is not always the one doing the manipulations; sometimes it’s a third party service like Imgix or Craft Cloud.

If there are additional modes/options you’d like added, feel free to request them individually, and we can consider those.
